File: ChatJetpackCompose/app/src/main/java/com/example/chatjetpackcompose/viewmodel/MesajlarViewModel.kt
package com.example.chatjetpackcompose.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.chatkotlin.model.Mesaj
import com.example.chatkotlin.service.RetrofitClient
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch

class MesajlarViewModel : ViewModel() {

    private val _mesajlar = MutableStateFlow<List<Mesaj>>(emptyList())
    val mesajlar: StateFlow<List<Mesaj>> = _mesajlar.asStateFlow()

    private val _hataMesaji = MutableStateFlow<String?>(null)
    val hataMesaji: StateFlow<String?> = _hataMesaji.asStateFlow()

    private val api = RetrofitClient.apiService
    private var mesajGuncelleJob: Job? = null

    fun mesajlariYuklePeriyodik(gonderenId: Int, aliciId: Int) {
        mesajGuncelleJob?.cancel()

        mesajGuncelleJob = viewModelScope.launch {
            while (isActive) {
                yukleMesajlar(gonderenId, aliciId)
                delay(15_000)
            }
        }
    }

    private suspend fun yukleMesajlar(gonderenId: Int, aliciId: Int) {
        try {
            val response = api.mesajlariGetir(gonderenId, aliciId)
            if (response.success) {
                _mesajlar.value = response.mesajlar
                _hataMesaji.value = null
            } else {
                _hataMesaji.value = "Mesajlar yüklenemedi"
            }
        } catch (e: Exception) {
            _hataMesaji.value = "Hata: ${e.message}"
        }
    }

    fun mesajGonder(
        gonderenId: Int,
        aliciId: Int,
        mesajText: String,
        base64Image: String? = null
    ) {
        val resimVar = if (base64Image.isNullOrEmpty()) 0 else 1

        viewModelScope.launch {
            try {
                val response = api.mesajGonder(
                    gonderenId,
                    aliciId,
                    mesajText,
                    resimVar,
                    base64Image
                )

                if (response.success) {
                    yukleMesajlar(gonderenId, aliciId)
                } else {
                    _hataMesaji.value = response.error ?: "Mesaj gönderilemedi"
                }

            } catch (e: Exception) {
                _hataMesaji.value = "Hata: ${e.message}"
            }
        }
    }

    override fun onCleared() {
        super.onCleared()
        mesajGuncelleJob?.cancel()
    }
}

File: flutter_chat_uygulama/lib/viewmodel/mesajlar_viewmodel.dart
import 'dart:async';

import 'package:flutter/material.dart';

import '../model/mesaj.dart';
import '../service/api_service.dart';

class MesajlarViewModel extends ChangeNotifier {
  final ApiService _api;

  MesajlarViewModel({ApiService? api}) : _api = api ?? ApiService();

  List<Mesaj> _mesajlar = [];
  List<Mesaj> get mesajlar => _mesajlar;

  String? _hataMesaji;
  String? get hataMesaji => _hataMesaji;

  Timer? _mesajGuncelleTimer;

  /// Kotlin: mesajlariYuklePeriyodik(gonderenId, aliciId)
  void mesajlariYuklePeriyodik({
    required int gonderenId,
    required int aliciId,
  }) {
    _mesajGuncelleTimer?.cancel();

    // İlk yükleme hemen
    _yukleMesajlar(gonderenId: gonderenId, aliciId: aliciId);

    _mesajGuncelleTimer = Timer.periodic(const Duration(seconds: 15), (_) {
      _yukleMesajlar(gonderenId: gonderenId, aliciId: aliciId);
    });
  }

  Future<void> _yukleMesajlar({
    required int gonderenId,
    required int aliciId,
  }) async {
    try {
      final response = await _api.mesajlariGetir(
        gonderenId: gonderenId,
        aliciId: aliciId,
      );

      if (response.success) {
        _mesajlar = response.mesajlar;
        _hataMesaji = null;
      } else {
        _hataMesaji = 'Mesajlar yüklenemedi';
      }
    } catch (e) {
      _hataMesaji = 'Hata: $e';
    }
    notifyListeners();
  }

  /// Kotlin: mesajGonder(...)
  Future<void> mesajGonder({
    required int gonderenId,
    required int aliciId,
    required String mesajText,
    String? base64Image,
  }) async {
    final resimVar = (base64Image == null || base64Image.isEmpty) ? 0 : 1;

    try {
      final response = await _api.mesajGonder(
        gonderenId: gonderenId,
        aliciId: aliciId,
        mesajText: mesajText,
        resimVar: resimVar,
        base64Img: base64Image,
      );

      if (response.success) {
        await _yukleMesajlar(gonderenId: gonderenId, aliciId: aliciId);
      } else {
        _hataMesaji = response.error ?? 'Mesaj gönderilemedi';
        notifyListeners();
      }
    } catch (e) {
      _hataMesaji = 'Hata: $e';
      notifyListeners();
    }
  }

  void stop() {
    _mesajGuncelleTimer?.cancel();
    _mesajGuncelleTimer = null;
  }

  @override
  void dispose() {
    stop();
    super.dispose();
  }
}
